src/actions/apply_attack_action.rs

```diff
@@ -76,6 +76,64 @@ fn binomial_coefficient(n: usize, k: usize) -> usize {
     result
 }
 
+fn gholdengo_scintillating_surfing(
+    acting_player: usize,
+    state: &State,
+) -> (Probabilities, Mutations) {
+    let active_pokemon = state.get_active(acting_player);
+    let metal_energy_count = active_pokemon
+        .attached_energy
+        .iter()
+        .filter(|&energy| *energy == EnergyType::Metal)
+        .count();
+
+    if metal_energy_count == 0 {
+        // No metal energy attached, no coins to flip
+        return probabilistic_damage_attack(vec![1.0], vec![0]);
+    }
+
+    // Generate all possible outcomes for flipping N coins
+    let num_outcomes = 2_usize.pow(metal_energy_count as u32);
+    let mut probabilities = vec![0.0; metal_energy_count + 1]; // 0 to metal_energy_count heads
+    let mut damages = Vec::new();
+
+    // For each possible outcome (0 to metal_energy_count heads)
+    for (heads, prob) in probabilities.iter_mut().enumerate() {
+        // Probability of getting exactly 'heads' heads out of 'metal_energy_count' coins
+        *prob = binomial_coefficient(metal_energy_count, heads) as f64 / (num_outcomes as f64);
+        damages.push((heads as u32) * 50); // 50 damage per heads
+    }
+
+    probabilistic_damage_attack(probabilities, damages)
+}
+
+fn croagunk_group_beatdown(
+    acting_player: usize,
+    state: &State,
+    damage_per_heads: u32,
+) -> (Probabilities, Mutations) {
+    let total_pokemon = state.enumerate_in_play_pokemon(acting_player).count();
+
+    if total_pokemon == 0 {
+        // No Pokémon in play, no coins to flip
+        return probabilistic_damage_attack(vec![1.0], vec![0]);
+    }
+
+    // Generate all possible outcomes for flipping N coins
+    let num_outcomes = 2_usize.pow(total_pokemon as u32);
+    let mut probabilities = vec![0.0; total_pokemon + 1]; // 0 to total_pokemon heads
+    let mut damages = Vec::new();
+
+    // For each possible outcome (0 to total_pokemon heads)
+    for (heads, prob) in probabilities.iter_mut().enumerate() {
+        // Probability of getting exactly 'heads' heads out of 'total_pokemon' coins
+        *prob = binomial_coefficient(total_pokemon, heads) as f64 / (num_outcomes as f64);
+        damages.push((heads as u32) * damage_per_heads);
+    }
+
+    probabilistic_damage_attack(probabilities, damages)
+}
+
```

Comment:
Same here can you re-use the logic? To keep the code DRY. Maybe a function that given some number of "counts" flips those many coins?